<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Rarefaction of reads samples — normalize_reads • standart</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Rarefaction of reads samples — normalize_reads"><meta name="description" content="Rarefaction of fasq files by sub-sampling the reads
before de novo assembly or alignment.
The normalization/standardization/sample size correction step allows to check
if some statistics are increasing with read numbers (e.g. heterozygous markers).
It's a very easy way to disentangle artifact from biological signal caused by
varying read numbers across samples."><meta property="og:description" content="Rarefaction of fasq files by sub-sampling the reads
before de novo assembly or alignment.
The normalization/standardization/sample size correction step allows to check
if some statistics are increasing with read numbers (e.g. heterozygous markers).
It's a very easy way to disentangle artifact from biological signal caused by
varying read numbers across samples."><meta property="og:image" content="https://thierrygosselin.github.io/standart/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">standart</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Installation</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-get-started" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Get started</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-get-started"><li><a class="dropdown-item" href="../articles/using_standart.html">Vignettes</a></li>
    <li><a class="dropdown-item" href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/thierrygosselin/standart" aria-label="GitHub"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Rarefaction of reads samples</h1>
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/standart/blob/main/R/normalize_reads.R" class="external-link"><code>R/normalize_reads.R</code></a></small>
      <div class="d-none name"><code>normalize_reads.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Rarefaction of fasq files by sub-sampling the reads
before de novo assembly or alignment.
The normalization/standardization/sample size correction step allows to check
if some statistics are increasing with read numbers (e.g. heterozygous markers).
It's a very easy way to disentangle artifact from biological signal caused by
varying read numbers across samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">normalize_reads</span><span class="op">(</span></span>
<span>  project.info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">fq.files</span>,</span>
<span>  sample.reads <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  number.replicates <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  random.seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-project-info">project.info<a class="anchor" aria-label="anchor" href="#arg-project-info"></a></dt>
<dd><p>(character, path, optional) When using the stackr pipeline,
a project info file is created. This file provides all the info and stats
generated by stacks and stackr.
The project info file will be updated with the new samples.
The project info filename will be appended <code>_normalize</code>.
The file should end with <code>.tsv</code>.
If no <code>project.info</code> file is provided, the function will have to look
at the number of reads in the fastq files and this will take longer.
Default: <code>project.info = NULL</code>.</p></dd>


<dt id="arg-fq-files">fq.files<a class="anchor" aria-label="anchor" href="#arg-fq-files"></a></dt>
<dd><p>(character, path) Path of folder containing the
samples to normalize.</p></dd>


<dt id="arg-sample-reads">sample.reads<a class="anchor" aria-label="anchor" href="#arg-sample-reads"></a></dt>
<dd><p>(integer) The number of reads to pick randomly.
Default: <code>sample.reads = 1000000</code>.</p></dd>


<dt id="arg-number-replicates">number.replicates<a class="anchor" aria-label="anchor" href="#arg-number-replicates"></a></dt>
<dd><p>(interger) The number of samples to generate.
With default, if 20 samples are in the folder, 100 new samples will be generated.
Default: <code>number.replicates = 5</code>.</p></dd>


<dt id="arg-random-seed">random.seed<a class="anchor" aria-label="anchor" href="#arg-random-seed"></a></dt>
<dd><p>(integer, optional) For reproducibility, set an integer
that will be used inside function that requires randomness. With default,
a random number is generated and printed in the appropriate output.
Default: <code>random.seed = NULL</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core for parallel
programming. Each samples to normalize is sequentially treated and replicates
are generated in parallel.
By default, <code>parallel.core = parallel::detectCores() - 1</code>.
This number is adjusted automatically to the number of replicates.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>fastq files with "-1", "-2", "..." appended to the original name.
If a project info file was provided, the new replicate samples info is integrated
to the file. The modified project info file will have <code>_normalize</code> appended
to the original filename.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thierrygosselin.github.io/standart/">standart</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># To run this function, bioconductor \code{ShortRead} package is necessary:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"http://bioconductor.org/biocLite.R"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">biocLite</span><span class="op">(</span><span class="st">"ShortRead"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Using OpenMP threads</span></span></span>
<span class="r-in"><span><span class="va">nthreads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/CallExternal.html" class="external-link">.Call</a></span><span class="op">(</span><span class="fu">ShortRead</span><span class="fu">:::</span><span class="va">.set_omp_threads</span>, <span class="fl">1L</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/on.exit.html" class="external-link">on.exit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/CallExternal.html" class="external-link">.Call</a></span><span class="op">(</span><span class="fu">ShortRead</span><span class="fu">:::</span><span class="va">.set_omp_threads</span>, <span class="va">nthreads</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># using defaults:</span></span></span>
<span class="r-in"><span><span class="fu">standart</span><span class="fu">::</span><span class="fu">normalize_reads</span><span class="op">(</span>fq.files <span class="op">=</span> <span class="st">"~/corals"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># customizing the function:</span></span></span>
<span class="r-in"><span><span class="fu">standart</span><span class="fu">::</span><span class="fu">normalize_reads</span><span class="op">(</span></span></span>
<span class="r-in"><span>   project.info <span class="op">=</span> <span class="st">"project.info.corals.tsv"</span>,</span></span>
<span class="r-in"><span>   fq.files <span class="op">=</span> <span class="st">"~/corals"</span>,</span></span>
<span class="r-in"><span>   sample.reads <span class="op">=</span> <span class="fl">2000000</span>,</span></span>
<span class="r-in"><span>   number.replicates <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>   random.seed <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>   parallel.core <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># You then need to run stackr: run_ustacks, run_sstacks, run_tsv2bam, run_gstacks, run_populations</span></span></span>
<span class="r-in"><span><span class="co"># or equivalent if a reference genome.</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

