<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Counts the number of reads in samples — read_counter • standart</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Counts the number of reads in samples — read_counter"><meta name="description" content="This function counts the number of reads in samples present
in the specified folder.
Useful if you don't have the info (e.g. generated by stacks process_radtags),
and you want to check the distribution in the number of reads between samples."><meta property="og:description" content="This function counts the number of reads in samples present
in the specified folder.
Useful if you don't have the info (e.g. generated by stacks process_radtags),
and you want to check the distribution in the number of reads between samples."><meta property="og:image" content="https://thierrygosselin.github.io/standart/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">standart</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/using_standart.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Computer setup</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/thierrygosselin/standart" aria-label="GitHub"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Counts the number of reads in samples</h1>
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/standart/blob/main/R/read_counter.R" class="external-link"><code>R/read_counter.R</code></a></small>
      <div class="d-none name"><code>read_counter.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function counts the number of reads in samples present
in the specified folder.
Useful if you don't have the info (e.g. generated by stacks process_radtags),
and you want to check the distribution in the number of reads between samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_counter</span><span class="op">(</span></span>
<span>  <span class="va">fq.files</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fq-files">fq.files<a class="anchor" aria-label="anchor" href="#arg-fq-files"></a></dt>
<dd><p>(character, path) Path of folder containing the
1 or more samples to count reads.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(optional)
The strata file is a tab delimited file with 2 columns headers:
<code>INDIVIDUALS</code> and <code>STRATA</code>.
The <code>STRATA</code> column can be any hierarchical grouping.
To create a strata file see <a href="https://thierrygosselin.github.io/radiator/reference/individuals2strata.html" class="external-link">individuals2strata</a>.
If you have already run
<a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">stacks</a> on your data,
the strata file is similar to a stacks <em>population map file</em>, make sure you
have the required column names (<code>INDIVIDUALS</code> and <code>STRATA</code>).
Note: Make sure that the fastq file names (without extension) match the INDIVIDUALS
column in the strata file. With default, figures are generated without strata
grouping.
Default: <code>strata = NULL</code>.</p></dd>


<dt id="arg-plot-reads">plot.reads<a class="anchor" aria-label="anchor" href="#arg-plot-reads"></a></dt>
<dd><p>With default <code>plot.reads = TRUE</code>, the distribution and
boxplot figures are generated and written in the directory.</p></dd>


<dt id="arg-write">write<a class="anchor" aria-label="anchor" href="#arg-write"></a></dt>
<dd><p>With default <code>write = TRUE</code>, the data frame with read counts
and figures are is written in the working directory.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core for parallel computing.
By default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list with a data frame with the sample id and the number of reads.
If option to generate figures was selected, the list also returns 2 figures
(see example below)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thierrygosselin.github.io/standart/">standart</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># To run this function, bioconductor \code{ShortRead} package is necessary:</span></span></span>
<span class="r-in"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"ShortRead"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># with defaults</span></span></span>
<span class="r-in"><span><span class="va">read.info</span> <span class="op">&lt;-</span> <span class="fu">standart</span><span class="fu">::</span><span class="fu">read_counter</span><span class="op">(</span>fq.files <span class="op">=</span> <span class="st">"corals"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># to extract info from the list</span></span></span>
<span class="r-in"><span><span class="va">reads</span> <span class="op">=</span> <span class="va">read.info</span><span class="op">$</span><span class="va">reads</span></span></span>
<span class="r-in"><span><span class="va">reads.distribution</span> <span class="op">&lt;-</span> <span class="va">read.info</span><span class="op">$</span><span class="va">reads.distribution</span></span></span>
<span class="r-in"><span><span class="va">reads.boxplot</span> <span class="op">&lt;-</span> <span class="va">read.info</span><span class="op">$</span><span class="va">reads.boxplot</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># If the default figures saved were not good, save with new width and height</span></span></span>
<span class="r-in"><span><span class="co"># the histogram</span></span></span>
<span class="r-in"><span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>filename <span class="op">=</span> <span class="st">"reads.distribution.pdf"</span>,</span></span>
<span class="r-in"><span>plot <span class="op">=</span> <span class="va">reads.distribution</span>,</span></span>
<span class="r-in"><span>width <span class="op">=</span> <span class="fl">15</span>, height <span class="op">=</span> <span class="fl">15</span>,</span></span>
<span class="r-in"><span>dpi <span class="op">=</span> <span class="fl">600</span>, units <span class="op">=</span> <span class="st">"cm"</span>, useDingbats <span class="op">=</span> <span class="cn">FALSE</span>, limitsize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># the boxplot</span></span></span>
<span class="r-in"><span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>filename <span class="op">=</span> <span class="st">"reads.boxplot.pdf"</span>,</span></span>
<span class="r-in"><span>plot <span class="op">=</span> <span class="va">reads.boxplot</span>,</span></span>
<span class="r-in"><span>width <span class="op">=</span> <span class="fl">15</span>, height <span class="op">=</span> <span class="fl">15</span>,</span></span>
<span class="r-in"><span>dpi <span class="op">=</span> <span class="fl">600</span>, units <span class="op">=</span> <span class="st">"cm"</span>, useDingbats <span class="op">=</span> <span class="cn">FALSE</span>, limitsize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

