<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using standart to inspect your reduced genome sequences • standart</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using standart to inspect your reduced genome sequences">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">standart</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/using_standart.html">Using standart to inspect your reduced genome sequences</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/standart/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using standart to inspect your reduced genome
sequences</h1>
                        <h4 data-toc-skip class="author">Thierry
Gosselin</h4>
            
            <h4 data-toc-skip class="date">2025-03-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/standart/blob/HEAD/vignettes/using_standart.Rmd" class="external-link"><code>vignettes/using_standart.Rmd</code></a></small>
      <div class="hidden name"><code>using_standart.Rmd</code></div>

    </div>

    
    
<p><img src="logo.png" align="right" height="100" alt=""></p>
<p>The vignette purpose is to show users how to use standart, an R
package that provides R functions to help solve the most common
challenges of genome reduction sequences quality control (DArT, RADseq,
GBS, etc).</p>
<p>standart really shines in projects that requires consistency among
sequencing runs and when samples are from animals with large and
complicated genomes.</p>
<div class="section level2">
<h2 id="coding-style">Coding style<a class="anchor" aria-label="anchor" href="#coding-style"></a>
</h2>
<ul>
<li>standart is trying hard to follow the <a href="https://style.tidyverse.org" class="external-link">tidyverse style guide</a>.</li>
<li>Inside stardart, using parallel execution is always through the use
of the argument: <code>parallel.core</code>.</li>
</ul>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<ul>
<li>Make sure you have the latest R and RStudio version before
installing stardart. My other package <a href="https://thierrygosselin.github.io/radiator/index.html" class="external-link">radiator</a>
as a nice <a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html" class="external-link">vignette</a>
for this.</li>
<li>Install stardart by following the <a href="https://github.com/thierrygosselin/standart#installation" class="external-link">instructions</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>I highly recommend running the sequencing files (entire lane or chip)
through a software dedicated for QC. If you have bad sequencing, no need
to go further in the pipeline…</p>
<ul>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>.</li>
<li>
<a href="http://www.sthda.com/english/wiki/fastqcr-an-r-package-facilitating-quality-controls-of-sequencing-data-for-large-numbers-of-samples" class="external-link">fastqcr</a>
is the R alternative.</li>
</ul>
<div class="section level4">
<h4 id="reads-length">reads length<a class="anchor" aria-label="anchor" href="#reads-length"></a>
</h4>
<p>If you are using Illumina sequencing, chances are that you have 100
pb reads (SE RADseq). With Ion Torrent, the size will be highly variable
and you have to determine the optimal size to cut your reads to the same
length, if the pipeline you are using for assembly requires it. Note
that most pipeline that performs <em>de novo</em> assembly works better
with uniform read length.</p>
<ol style="list-style-type: decimal">
<li>output of <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
</li>
</ol>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
will output the read length distribution. But it’s difficult to know how
many markers you will get depending on read size.</p>
<p>You will see the relationship between the position of the nucleotide
on the read and overall quality. This is very helpful to determine if
you want to cut all your Illumina reads to 80 or 90 pb.</p>
<p><strong>Below is a FastQC figure of a typical Ion Torrent sequencing
chip sequence length distribution:</strong>
<img src="ion_torrent_1.png"></p>
<ol start="2" style="list-style-type: decimal">
<li>Use standart</li>
</ol>
<p>Below is a figure generated by standart that shows the number of
reads you will get depending on the maximum size threshold selected:</p>
<p><img src="ion_torrent_2.png"></p>
<p>The function is:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fu">standart</span><span class="fu">::</span><span class="fu"><a href="../reference/reads_length_distribution.html">reads_length_distribution</a></span><span class="op">(</span></span>
<span>  fq.file <span class="op">=</span> <span class="st">"my-fastq.fastq.gz"</span>, </span>
<span>  with.future <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12L</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>You can use individual file or entire chip or lane. Function
documentation: <code><a href="../reference/reads_length_distribution.html">?standart::reads_length_distribution</a></code></p>
<p><strong>Remember that depending of types of file (individual or
entire lane/chip:</strong></p>
<ul>
<li>Barcodes and adapters might be in this relationship observed in the
figure.</li>
<li>It’s not uncommon to see increasing adapters present with increasing
read size.</li>
<li>Repetitive sequences, unique reads and so on might impact the
results (<a href="#noise">see noise section below</a>).</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="noise">Noise reduction<a class="anchor" aria-label="anchor" href="#noise"></a>
</h3>
<p>Reducing the noise in an individual FQ file could really help
speeding up your analysis. I use it when I have more than 200 samples.
Nothing new here:</p>
<ul>
<li>Daniel Ilut explains it in his pipeline (Ilut et al. 2014) and my
function here his basically the same as his but in R…</li>
<li>Jon Puritz’s <a href="http://www.ddocent.com/UserGuide/#de-novo-assembly" class="external-link">dDocent data
reduction step</a> as a similar example.</li>
<li>Praveen’s pipeline <a href="https://github.com/beiko-lab/RADProc" class="external-link">RADProc</a> (Ravindran et
al. 2019) use something similar.</li>
<li>Metagenomic analysis pipeline all have some step of denoising the
data.</li>
</ul>
<div class="section level4">
<h4 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h4>
<p>I prefer Daniel Ilut way of visualizing the information.</p>
<p>The function <code>stardart::read_depth_plot</code> requires 1 FQ
file (after stacks <em>process_radtags</em> or
<code><a href="http://thierrygosselin.github.io/stackr/reference/run_process_radtags.html" class="external-link">stackr::run_process_radtags</a></code>) and <strong>is very fast to
run</strong>. It will highlight read coverage groups of distinct reads
within a sample. When I say distinct reads, is reads that are exactly
the same, no mismatches. The distinct reads are putative allele.</p>
<p><img src="read_depth_coverage_groups.png"></p>
<p><strong>4 read coverage groups are shown:</strong></p>
<ul>
<li><p><strong>distinct reads with low coverage (in red): </strong>
these reads are likely sequencing errors or uninformative polymorphisms
(shared only by a few samples).</p></li>
<li>
<p><strong>disting reads for a target coverage (in green):
</strong></p>
<ul>
<li>Usually represent around 80% of the reads in the FQ file.</li>
<li>It’s a <em>safe coverage range</em> to start exploring your data
(open for discussion).</li>
<li>Lower threshold (default = 7): you could go as low as maybe 3 ? You
can’t escape it, it’s your tolerance to call heterozygote <strong>a true
heterozygote</strong>. You want a minimum coverage for both the
reference and the alternative allele. Yes, you can use population
information to lower this threshold or use some fancy <em>bayesian</em>
algorithm.</li>
<li>Higher threshold: is a lot more open for discussion, here it’s the
lower limit of another group (the orange, see below for description).
Minus 1 bp.</li>
</ul>
</li>
<li><p><strong>distinct reads with high coverage &gt; 1 read depth (in
yellow): </strong> those are legitimate alleles with high
coverage.</p></li>
<li><p><strong>distinct and unique reads with high coverage (in orange):
</strong> this is where you’ll find paralogous sequences, transposable
elements and other high copy number variants.</p></li>
</ul>
<p>To run it on your data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">read.plot</span> <span class="op">&lt;-</span> <span class="fu">standart</span><span class="fu">::</span><span class="fu"><a href="../reference/read_depth_plot.html">read_depth_plot</a></span><span class="op">(</span></span>
<span>  fq.file <span class="op">=</span> <span class="st">"my-fq.fq.gz"</span>,</span>
<span>  min.coverage.fig <span class="op">=</span> <span class="fl">7L</span>, <span class="co">#default</span></span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span> <span class="co">#default</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The function will generate the plot shown above and write the figure
in the working directory, using the sample name:
<code>my-fq_hap_read_depth.png</code>.</p>
</div>
<div class="section level4">
<h4 id="faq">FAQ<a class="anchor" aria-label="anchor" href="#faq"></a>
</h4>
<p><strong>Should I run the function on all samples ?</strong></p>
<p>Please don’t. Their is no use in running this on the entire dataset.
It’s going to be fairly the same. Run it on a couple of samples. Based
on the number of reads (distribution), choose 1 sample: below average,
average and above average and compare the look of the figures
produced.</p>
</div>
<div class="section level4">
<h4 id="filtering">Filtering ?<a class="anchor" aria-label="anchor" href="#filtering"></a>
</h4>
<p>Removing the noise in FQ files, it’s usually a hot topic… Personally,
the less I do the better and it’s definitely the case for small
projects. I don’t think the noise is a problem for pipline like stacks.
For large projects, it’s something else, and certainly a huge
bottleneck. When you generate <em>de novo</em> assembly for more than
2000 samples and generate a catalog with all the noise, cleaning the
data can represent weeks of computation time that you shave off your
project!</p>
<p>To test:</p>
<p><strong>1. Clean the FQ file</strong></p>
<p>Use the sample you just used in the read coverage groups, above,
clean it with <code><a href="../reference/clean_fq.html">standart::clean_fq</a></code>. It’s very fast to
run:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clean.fq.name</span> <span class="op">&lt;-</span> <span class="fu">standart</span><span class="fu">::</span><span class="fu"><a href="../reference/clean_fq.html">clean_fq</a></span><span class="op">(</span></span>
<span>  fq.file <span class="op">=</span> <span class="st">"my-fq.fq.gz"</span>, </span>
<span>  min.coverage.threshold <span class="op">=</span> <span class="fl">2</span>, </span>
<span>  remove.unique.reads <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#default</span></span>
<span>  write.blacklist.fasta <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write.blacklist <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#default</span></span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>, <span class="co">#default</span></span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span> <span class="co">#default</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>This will generate:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my</span><span class="op">-</span><span class="va">fq_cleaned.fq.gz</span> <span class="co"># the cleaned fq file</span></span>
<span><span class="va">my</span><span class="op">-</span><span class="va">fq_blacklisted_reads.fasta.gz</span> <span class="co"># a fasta file with blacklisted sequences...</span></span></code></pre></div>
<p><strong>2. Run your <em>de novo</em> assembly pipeline</strong></p>
<p>Run your <em>de novo</em> assembly pipeline (e.g. stacks or stackr
and do the next step below: <em>the mismatch threshold series</em>
(using <code><a href="http://thierrygosselin.github.io/stackr/reference/run_ustacks.html" class="external-link">stackr::run_ustacks</a></code>). For this, use the raw and
clean FQ files and compare the output…</p>
<p>In all my projects, having less reads (the cleaned FQ file)
means:</p>
<ul>
<li>faster <em>de novo</em> assemblies (I’ve seen up to 3 times
faster).</li>
<li>less loci generated, but of better qualities:
<ul>
<li>higher coverage in all the <em>de novo</em> assembly steps of
stacks</li>
<li>less repetitive stacks blacklisted</li>
<li>lower number of locus with assembly artifacts (lors of SNPs/locus,
etc)</li>
<li>very similar heterozygosity and homozygosity proportions</li>
</ul>
</li>
<li>the mismathches plots are almost always identical: showing me that I
can use the same Mismatch threshold (the <code>-M</code> in stacks
<em>ustacks</em>) for the cleaned or the raw FQ files.</li>
<li>faster catalog generation time, the stacks <em>cstacks</em> or
<code><a href="http://thierrygosselin.github.io/stackr/reference/run_cstacks.html" class="external-link">stackr::run_cstacks</a></code> step.</li>
<li>NO DIFFERENCES in downstream analysis (population structure, outlier
detection, Fst, assignment, closekin).</li>
</ul>
<p><strong>Why keep the blacklisted sequences (the fasta file)
?</strong></p>
<p>You could use <code><a href="https://rdrr.io/pkg/Biostrings/man/matchPattern.html" class="external-link">Biostrings::countPattern</a></code> function to look
for TE in those sequences;)</p>
<p><strong>3. Filtering the entire dataset</strong></p>
<p>If you see a big difference and are convince that filtering is the
way to go, you can easily do it with this piece of code:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fq.to.clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"04_process_radtags/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fq.to.clean</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">fq.to.clean</span></span>
<span></span>
<span><span class="va">clean.all</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_df</a></span><span class="op">(</span></span>
<span>  .x <span class="op">=</span> <span class="va">fq.to.clean</span>, </span>
<span>  .f <span class="op">=</span> <span class="fu">standart</span><span class="fu">::</span><span class="va"><a href="../reference/clean_fq.html">clean_fq</a></span>,</span>
<span>  min.coverage.threshold <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  remove.unique.reads <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#default</span></span>
<span>  write.blacklist.fasta <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write.blacklist <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#default</span></span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Transfer the cleaned files or the raw ones in a separate
folder/sub-folder to keep things tidy.</p>
</div>
</div>
<div class="section level3">
<h3 id="normalization">Data normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>Another hot topic, up for debate…</p>
<p>After the stacks step called <em>process_radtags</em>, in stackr
<code>run_process_radtags</code>, looking at the distribution of the
number of reads per sample, you might be getting one of these
distribution:</p>
<p><img src="read_distribution.png"></p>
<p>In an ideal world, <strong>you want the uniform
distribution</strong>. That translate in good field and wet-lab
techniques:</p>
<ul>
<li>tissue sampling and preservation</li>
<li>extraction (robots ?)</li>
<li>pipetting</li>
<li>DNA quantification</li>
<li>all the other required techniques</li>
</ul>
<p>And this resulted in similar sequenced information between your
samples. You’re about to compare apples with apples. :)</p>
<p>If you have a skewed (left or right) or a normal distribution, you’ll
be comparing apples with … citrus. The samples with lower reads will
have less loci and usually less heterozygous loci. Similarly, the
samples with higher number of reads will have more loci, more
heterozygous loci.</p>
<p>What’s the likely cause for these ?</p>
<ul>
<li>student with no prior wet-lab experience doing the DNA extraction or
downstream step (e.g. library prep)</li>
<li>data preservation issues: historical DNA, fragmented DNA, etc.</li>
<li>skipping the DNA quantification or library QC steps</li>
</ul>
<p>Ok, but what’s the problem with that ?</p>
<p><strong>Potentially these problems:</strong></p>
<ul>
<li>ascertainment bias that could drive populations polymorphism
discovery bias</li>
<li>missing data patterns</li>
<li>individual heterozygosity problems or patterns</li>
<li>downstream analysis impacted by the differences between your
samples: artifactual or biological ?</li>
</ul>
<p><strong>Solutions ?</strong></p>
<div class="section level4">
<h4 id="wet-lab-normalization">1. Wet-lab normalization<a class="anchor" aria-label="anchor" href="#wet-lab-normalization"></a>
</h4>
<p>Here are the basic steps:</p>
<ul>
<li>sequence the samples</li>
<li>look at the stats we’ve discussed</li>
<li>go back in the wet-lab to prepare a new plate</li>
<li>sequence</li>
<li>pool or not the lanes/chips</li>
<li>look at the sequencing statistics</li>
<li>repeat the step above or continue with the pipeline</li>
</ul>
<p>I’m not a fan:</p>
<ul>
<li>Way to many wet-lab steps.</li>
<li>The wet-lab is <strong>THE MOST</strong> important part.</li>
<li>Everything gets exacerbated with RADseq.</li>
<li>I’m not in favor of pooling lanes or chips (more on this some
day!).</li>
<li>It’s really not like working with microsatelites.</li>
</ul>
</div>
<div class="section level4">
<h4 id="bio-informatic-normalization">2. Bio-informatic normalization<a class="anchor" aria-label="anchor" href="#bio-informatic-normalization"></a>
</h4>
<p>When doing it right from the start is impossible, I prefer to do the
normalization after de demultiplexing step… after for example
<code><a href="http://thierrygosselin.github.io/stackr/reference/run_process_radtags.html" class="external-link">stackr::run_process_radtags</a></code>. The prerequisite is to have
enough sequencing material (reads) for all my samples.</p>
<p>The tool? <a href="http://thierrygosselin.github.io/standart/reference/normalize_reads.html" class="external-link">standart::normalize_reads</a>
function.</p>
<p>Rarefaction of fasq files by sub-sampling the reads before <em>de
novo</em> assembly or alignment is not new:</p>
<ul>
<li>
<strong>Community and biodiversity ecologists</strong> are well
aware of species accumulation curves obtained with rarefaction (Sanders,
1968). The rarefaction method guarantees the samples have the same
weight in the analysis.</li>
<li>
<strong>Metagenomics community</strong> before analyzing the alpha
diversity will deal with differing sample depth for a given diversity
statistic by using rarefaction. It means taking a random subset of a
given size of the original sample (see <a href="https://mothur.org" class="external-link">mothur</a>, Schloss et al. 2009).</li>
<li>
<strong>Molecular ecologists</strong> for some reasons, RADseq
enthusiasts seems unaware of this concept, but see (Hale et al. 2009;
Kalinowski, 2004 and Puncher et al. 2018).</li>
<li>
<strong>The Keywords:</strong> rarefaction, sub-sampling,
normalization, standardization, sample size correction.</li>
</ul>
<p><strong>The steps ?</strong></p>
<ul>
<li>the goal is to reduce variance in number of reads between
samples</li>
<li>rarefaction/normalization is conducted aon the fq file after stacks
<em>process_radtags</em> and after removing the noise.</li>
<li>normalized replicates are generated by the function by randomly
selecting the reads (without replacement) for each samples</li>
<li>you choose the number of replicates (default: 3) and you manage
those replicates like other replicates you might have (you want
replicates…).</li>
</ul>
<p>Read the <a href="http://thierrygosselin.github.io/stackr/reference/normalize_reads.html#examples" class="external-link">function
documentation and look at the example</a></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To run this function, bioconductor's ShortRead package is necessary:</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ShortRead"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using defaults:</span></span>
<span><span class="fu">standart</span><span class="fu">::</span><span class="fu"><a href="../reference/normalize_reads.html">normalize_reads</a></span><span class="op">(</span>path.samples <span class="op">=</span> <span class="st">"~/corals"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># customizing the function:</span></span>
<span><span class="fu">standart</span><span class="fu">::</span><span class="fu"><a href="../reference/normalize_reads.html">normalize_reads</a></span><span class="op">(</span></span>
<span>   project.info <span class="op">=</span> <span class="st">"project.info.corals.tsv"</span>,</span>
<span>   path.samples <span class="op">=</span> <span class="st">"04_process_radtags"</span>,</span>
<span>   sample.reads <span class="op">=</span> <span class="fl">2000000</span>,</span>
<span>   number.replicates <span class="op">=</span> <span class="fl">3</span>,</span>
<span>   random.seed <span class="op">=</span> <span class="fl">3</span>,</span>
<span>   parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span>   <span class="op">)</span></span></code></pre></div>
<p><strong>My observations using rarefaction:</strong></p>
<ul>
<li>the replicates have similar mismathches plots: showing me that I can
use the same Mismatch threshold (the <code>-M</code> in stacks
<em>ustacks</em>) for the rarefied samples.</li>
<li>better <em>de novo</em> assembly statistics</li>
<li>when the individual or population heterozygosity <strong>is a real
biological signal</strong>, rarefaction <strong>doesn’t impact the
downstream analysis</strong>.</li>
<li>when the heterozygosity was driven by the sequencing bias,
bioinformatic rarefaction took care of the problem.</li>
</ul>
<div class="section level5">
<h5 id="faq-1">FAQ<a class="anchor" aria-label="anchor" href="#faq-1"></a>
</h5>
<p><strong>It doesn’t bother you to throw away all the valuable
information that samples with high read numbers have ?</strong></p>
<p>Absolutely not!</p>
<ul>
<li>
<strong>most of the information is already lost</strong>: that is if
you used stacks to generate a catalog and then filtered the data to keep
markers in x individuals and/or in all sampling sites/populations. At
this step it’s either you loose markers or get more missing data…</li>
<li>
<strong>It’s not a loss of information if the information cannot be
confirmed with other samples:</strong> it’s a statistic problem and
ecologists are well aware of this.</li>
<li>
<strong>The loss of reads is intrinsic to RADseq: </strong> Don’t
believe me, inspect stacks, stackr or any RADseq pipeline log files…
Rarefaction is just another step, molecular ecologists are not used to
it.</li>
<li>check the total read depth of 1 sample after all the bioinformatic
steps. Compare the numbers with the actual read number from the start of
the pipeline. If you have &gt; 30% you are lucky!</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Ilut, D., Nydam, M., Hare, M. (2014). Defining Loci in
Restriction-Based Reduced Representation Genomic Data from Non model
Species: Sources of Bias and Diagnostics for Optimal Clustering BioMed
Research International 2014(2), 1 9. <a href="https://dx.doi.org/10.1155/2014/675158" class="external-link uri">https://dx.doi.org/10.1155/2014/675158</a></p>
<p>Rochette, N., Rivera‐Colón, A., Catchen, J. (2019). Stacks 2:
Analytical methods for paired‐end sequencing improve RADseq‐based
population genomics. Molecular Ecology <a href="https://dx.doi.org/10.1111/mec.15253" class="external-link uri">https://dx.doi.org/10.1111/mec.15253</a></p>
<p>Harvey, M., Judy, C., Seeholzer, G., Maley, J. (2015). Similarity
thresholds used in short read assembly reduce the comparability of
population histories across species PeerJ 3(), e895. <a href="https://dx.doi.org/10.7717/peerj.895" class="external-link uri">https://dx.doi.org/10.7717/peerj.895</a></p>
<p>Ravindran, P., Bentzen, P., Bradbury, I., Beiko, R. (2019). RADProc:
A computationally efficient de novo locus assembler for population
studies using RADseq data Molecular Ecology Resources 19(1), 272-282. <a href="https://dx.doi.org/10.1111/1755-0998.12954" class="external-link uri">https://dx.doi.org/10.1111/1755-0998.12954</a></p>
<p>Schloss, P.D., et al., Introducing mothur: Open-source,
platform-independent, community-supported software for describing and
comparing microbial communities. Appl Environ Microbiol, 2009.
75(23):7537-41.</p>
<p>Sanders HL (1968) Marine benthic diversity: A comparative study. The
American Naturalist 102: pp. 243–282</p>
<p>Hale et al. (2009) on the relative merits of normalization and
rarefaction in gene discovery in sturgeons. BMC genomics, 10, 203.</p>
<p>Kalinowski (2004) Counting Alleles with Rarefaction. Conservation
Genetics, 5, 539–543.</p>
<p>Puncher et al. (2018) Molecular Ecology Resources, 44, 678.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
